### PR TITLE
emscripten: 3.1.17 -> 3.1.24

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

It moves to `buildNpmPackage` as there is no `yarn.lock` anymore on upstream, note that it uses lockfile v1 which produces big trace regarding lockfile versions.

https://github.com/emscripten-core/emscripten/blob/main/ChangeLog.md

It cannot be bumped further without #194634.

For ≥ 3.1.27 (IIRC), binaryen must be bumped too.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/emscripten/default.nix [hunk 1]
```diff
@@ -2,31 +2,39 @@
...
-    packageJSON = ./package.json;
-    yarnLock = ./yarn.lock;
-    yarnNix = ./yarn.nix;
+    inherit pname version src;
+
+    npmDepsHash = "sha256-ejuHR2BpAUStWjuvQuGE6ko4byF4GBl6FJBshxlknQk=";
+
+    # Do not build.
+    buildPhase = ''
+      echo no build
```
Comment (RaitoBezarius) [inline pkgs/development/compilers/emscripten/default.nix:25]: I guess `dontNpmBuild = true; dontBuild = true;` would work here?

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 206005` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2jack</li>
    <li>faust2jaqt</li>
    <li>faust2lv2</li>
    <li>open-music-kontrollers.mephisto</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>emscripten</li>
    <li>faust (faust2)</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jackrust</li>
    <li>faust2ladspa</li>
  </ul>
</details>

**Comment by willcohen** [comment]
Given the binaryen and llvm issues mentioned in the issue, this looks good to me!

**Review by willcohen — APPROVED** [review]

**Comment by willcohen** [comment]
Is there anything left to do here before merging?

**Comment by willcohen** [comment]
@RaitoBezarius in the meantime, since you're probably the person to most recently work on bumping this:

I've been playing a little bit with trying to build on top of this and the llvm-15 branch to see if I can get current emscripten to build. https://github.com/willcohen/nixpkgs/commits/emscripten-llvm-15

There's a few stricter warnings that need to be disabled, and the patch has to get rebased, but it's mostly working. The main hurdle I'm currently hitting is this:

```
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libsockets.a... (this will be cached in "/nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libsockets.a" for subsequent builds)
system_libs:INFO: compiled 21 inputs
cache:INFO:  - ok
wasm-ld: error: /nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libcompiler_rt.a(stack_limits.o): undefined symbol: __stack_high
wasm-ld: error: /nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libcompiler_rt.a(stack_limits.o): undefined symbol: __stack_low
emcc: error: '/nix/store/6jdsg3qnvk3h47zilhfm0i7iqwlr4k3n-emscripten-llvm-3.1.29/bin/wasm-ld -o a.out.wasm /private/tmp/nix-build-emscripten-3.1.29.drv-2/emscripten_temp_amja185_/test_0.o -L/nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto /nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/crt1.o -lGL -lal -lhtml5 -lstandalonewasm -lstubs-debug -lc-debug -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-debug-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --strip-debug --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export-table -z stack-size=65536 --initial-memory=16777216 --max-memory=16777216 --stack-first' failed (returned 1)
```

This looks like it's related to https://github.com/emscripten-core/emscripten/commit/830fcb81c3b8a606157ece3dc72671d92947d3fa, which means it'll get hit for 3.1.25+. Any hunch on how to get those two things defined for the builds?

**Comment by RaitoBezarius** [comment]
> @RaitoBezarius in the meantime, since you're probably the person to most recently work on bumping this:
> 
> I've been playing a little bit with trying to build on top of this and the llvm-15 branch to see if I can get current emscripten to build. https://github.com/willcohen/nixpkgs/commits/emscripten-llvm-15
> 
> There's a few stricter warnings that need to be disabled, and the patch has to get rebased, but it's mostly working. The main hurdle I'm currently hitting is this:
> 
> ```
> cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libsockets.a... (this will be cached in "/nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libsockets.a" for subsequent builds)
> system_libs:INFO: compiled 21 inputs
> cache:INFO:  - ok
> wasm-ld: error: /nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libcompiler_rt.a(stack_limits.o): undefined symbol: __stack_high
> wasm-ld: error: /nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libcompiler_rt.a(stack_limits.o): undefined symbol: __stack_low
> emcc: error: '/nix/store/6jdsg3qnvk3h47zilhfm0i7iqwlr4k3n-emscripten-llvm-3.1.29/bin/wasm-ld -o a.out.wasm /private/tmp/nix-build-emscripten-3.1.29.drv-2/emscripten_temp_amja185_/test_0.o -L/nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto /nix/store/gj96xc8p1cybb4cr1sc72cp19h5qbvxx-emscripten-3.1.29/share/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/crt1.o -lGL -lal -lhtml5 -lstandalonewasm -lstubs-debug -lc-debug -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-debug-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --strip-debug --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export-table -z stack-size=65536 --initial-memory=16777216 --max-memory=16777216 --stack-first' failed (returned 1)
> ```
> 
> This looks like it's related to [emscripten-core/emscripten@830fcb8](https://github.com/emscripten-core/emscripten/commit/830fcb81c3b8a606157ece3dc72671d92947d3fa), which means it'll get hit for 3.1.25+. Any hunch on how to get those two things defined for the builds?

Hm, I supposed it was related to not having LLVM15, but you seem to be using LLVM15, so… I have no idea :(
But I have plan to come back and try to fix this when LLVM15 is stabilized

> Is there anything left to do here before merging?

AFAIK no, except merge conflicts stuff

File: pkgs/development/compilers/emscripten/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/emscripten/default.nix:25]: dontBuild is it probaby.

File: pkgs/development/compilers/emscripten/default.nix [hunk 2]
```diff
@@ -2,31 +2,39 @@
...
+    '';
+
+    # Copy node_modules directly.
+    installPhase = ''
+      cp -r node_modules $out/
+    '';
```
[the remark this answers is not in the document]
Comment (willcohen) [reply pkgs/development/compilers/emscripten/default.nix:31]: This actually becomes the output of the new emscripten-node-modules in nix store. I tried it using `ln -s` in #210397 but it wasn't able to create it at that level.

**Comment by willcohen** [comment]
Rebased this to master and adjusted to use `dontBuild` as suggested by @SuperSandro2000 in #210397.

**Comment by RaitoBezarius** [comment]
Superseded by #210397

**Pull request closed without merging**